Description: Support custom stream wrappers in XMLReader::open().
 This is similar to the treatment that was applied to a few other places in
 842d6fb7dcd6d: manually create an IO object and use that to create the
 xmlTextReader, rather than just calling xmlReaderForFile. This fixes web
 importing of pages in MediaWiki.
Author: Brett Simmers <bsimmers@fb.com>
Forwarded: Not needed
Last-Update: 2014-09-25

---
--- a/hphp/runtime/ext/ext_xmlreader.cpp
+++ b/hphp/runtime/ext/ext_xmlreader.cpp
@@ -127,7 +127,18 @@ bool c_XMLReader::t_open(const String& uri, const String& encoding /*= null_stri
   xmlTextReaderPtr reader = NULL;
 
   if (!valid_file.empty()) {
-    reader = xmlReaderForFile(valid_file.data(), encoding.data(), options);
+    // Manually create the IO context to support custom stream wrappers.
+    auto stream = File::Open(valid_file, "rb");
+    if (!stream.isInvalid()) {
+      reader = xmlReaderForIO(libxml_streams_IO_read,
+                              libxml_streams_IO_close,
+                              stream.get(),
+                              valid_file.data(),
+                              encoding.data(),
+                              options);
+      // The xmlTextReaderPtr owns a reference to stream.
+      if (reader) stream.get()->incRefCount();
+    }
   }
 
   if (reader == NULL) {
